### DISCUSSION TITLE
Gantt chart library for react: passing necessary options object to react wrapper

### DISCUSSION THREAD

**Discussion by philrwebdev** [neuefische/questions#847] [Web] [answered]
We would like to use a Gantt chart for our drop-plants app project to allow users to indicate to allow users to define periods of relevant tasks for specific crops (eg seed period). There are several libraries available, but from what I have seen so far, none works out of the box for our specific purpose. One requirement for our app is that the visible time period is exactly one calendar year (Jan-Dec) and no scrolling to previous or later time points extending beyond this calendar year is possible. Within this year, users should be able to define a time period.

Our favorite library for the implementation of this feature is frappe gantt: https://frappe.io/gantt (github: https://github.com/frappe/gantt/blob/master/README.md). This library allows to set the 'padding' to the left and right of the time period in view (see https://github.com/frappe/gantt/pull/387, https://github.com/frappe/gantt/pull/330, https://github.com/frappe/gantt/issues/169). I am not sure this will completely handle our case, since the chart may still be scrollable though. The padding can be set via 'view_mode_padding' in an options object passed.

There is a react wrapper for the JS library on github: https://github.com/Soremwar/react-frappe-gantt. As far as I can see, the component extending the Gantt class is defined here: https://github.com/Soremwar/react-frappe-gantt/blob/master/src/components/FrappeGantt.jsx. However, here I can't see how the options object required to customize (among other things) the padding can be passed, since the options object (third argument) seems to only take a subset of prespecified options:

```
componentDidMount() {
    this._gantt = new Gantt(this._svg.current, this.state.tasks, {
      on_click: this.props.onClick,
      on_view_change: this.props.onViewChange,
      on_progress_change: (task, progress) => {
        this.props.onProgressChange(task, progress);
        this.props.onTasksChange(this.props.tasks);
      },
      on_date_change: (task, start, end) => {
        this.props.onDateChange(task, start, end);
        this.props.onTasksChange(this.props.tasks);
      },
    });
```

Using the react-frappe-gantt wrapper works for me. But when I try to import the original Gantt function as done here (`import Gantt from "frappe-gantt";`); I get the following error message: `Module not found: Can't resolve 'frappe-gantt'`. Also, while importing the react-wrapper works, VS code marks the `import { FrappeGantt } from "react-frappe-gantt";` statement with a warning: `Could not find a declaration file for module 'react-frappe-gantt'`. 

Hence my question: Is there a way to (easily) pass the required options object to the `FrappeGantt` component defined in the react-wrapper (https://github.com/Soremwar/react-frappe-gantt/blob/master/src/components/FrappeGantt.jsx)? Maybe by adapting the react wrapper? If necessary, how can the original JS function be imported? If scrolling beyond the calendar year is still possible, is there a react-compatible way of blocking this?

Any hints welcome!

PS: Here is a minimal working example for the react-frappe-gantt wrapper (install with `npm install react-frappe-gantt`). Adding `view_mode_padding= {{MONTH: [[0, 'month'], [0, 'month']],}}` or similar (see https://github.com/frappe/gantt/pull/330) as a prop does not seem to change anything.

```
import { FrappeGantt } from "react-frappe-gantt";

export default function TestPage() {

  const tasks = [
    {
      id: "Task 1",
      name: "Redesign website",
      start: "2016-12-28",
      end: "2016-12-31",
      progress: 10,
      dependencies: "",
    },
    {
      id: "Task 2",
      name: "Redesign website",
      start: "2016-12-28",
      end: "2016-12-31",
      progress: 20,
      dependencies: "Task 1",
    },
    {
      id: "Task 3",
      name: "Redesign website",
      start: "2016-12-28",
      end: "2016-12-31",
      progress: 0,
      dependencies: "Task 2, Task 1",
    },
  ];

  return (
    <>
      <FrappeGantt
        onClick={(task) => console.log(task)}
        onDateChange={(task, start, end) => console.log(task, start, end)}
        onProgressChange={(task, progress) => console.log(task, progress)}
        onTasksChange={(tasks) => console.log(tasks)}
        tasks={tasks}
        viewMode="Month"
      />
    </>
  );
}
```

**Comment by doemser** [comment, marked as answer]
The group decided to implement a custom gant chart-type of input and display component
